Add package: git-review

healther opened the pull request.

adamjstewart commented on var/spack/repos/builtin/packages/git-review/package.py:
```diff
@@ -0,0 +1,52 @@
+    version('1.24',   '145116fe58a3487c3ad1bf55538fd741')
+    version('1.23',   'b0023ad8c037ab710da81412194c6a3a')
+    version('1.22',   'e889df5838c059362e5e0d411bde9c48')
+    version('1.21',   'eee88bdef1aa37a55cc8becd48c6aba9')
```
It doesn't seem possible to download an older version of this package since the URL cannot be extrapolated.

adamjstewart commented on var/spack/repos/builtin/packages/git-review/package.py:
```diff
@@ -0,0 +1,52 @@
+
+    extends('python')
+
+    depends_on('py-setuptools',    type=('build', 'run'))
```
Is this actually needed during run-time? For most packages, it is only a build dependency, although occasionally it is needed to run as well.

adamjstewart commented on var/spack/repos/builtin/packages/git-review/package.py:
```diff
@@ -0,0 +1,52 @@
+    depends_on('py-pbr',           type=('build', 'run'))
+    depends_on('py-requests@1.1:', type=('build', 'run'))
+    depends_on('git',              type=('build', 'run'))
+    depends_on('tk',               type=('build', 'run'))
```
This doesn't need to link to git or tk?

adamjstewart commented on var/spack/repos/builtin/packages/git-review/package.py:
```diff
@@ -0,0 +1,52 @@
+
+    def install(self, spec, prefix):
+        os.environ['PBR_VERSION'] = str(spec.version)
+        setup_py('install', '--prefix={0}'.format(prefix))
```
For this package, I would extend the PythonPackage base class and get rid of the install method. In order to set the PBR_VERSION, you can add a setup_environment method and add it to spack_env. I think py-rtree does something similar if you need an example.

adamjstewart commented:
> I'm dubious that 1.24 is a real version, wouldn't it be 1.24.5 or something?

It looks like that is a real version: https://pypi.python.org/pypi/git-review/1.24

healther commented:
I'll do that when I'm back

healther commented:
Rebased this branch onto the current develop head and added git-review dependencies (wasn't a good idea to split that up in the first place)

Regarding the url's: I kind of gave up trying to get pypi to work with older versions, but the github releases would be preferable anyways as they are automatically extendable, or am I missing something?

As a side note: I noticed that some github-urls fail to download with ```curl``` but work with ```wget```. Can I tell spack to prefer ```wget``` over ```curl```?

adamjstewart commented:
I generally try to use PyPI when I can. If you use `pypi.io` it isn't hard to extrapolate older releases. If you do decide to use GitHub, please make sure that you have a checksum for the version in the URL so my url parsing commands can tell that they're working correctly.

As for wget vs curl, it currently isn't configurable. If you use Spack to build a newer version of curl and add it to your PATH you'll find that it should work fine for those problematic URLs.

citibeth commented on var/spack/repos/builtin/packages/py-coverage/package.py:
```diff
@@ -29,8 +29,9 @@ class PyCoverage(PythonPackage):
 
+    version('4.3.4', 'd347766b06bbb4fd0bc822014b7cfb0a')
     version('4.0a6', '1bb4058062646148965bef0796b61efc')
 
```
It is important that URLs NOT be changed without good cause.  As @adamjstewart mentioned, we normally download Python stuff from PyPI; this recommendation is also supported by the Py-Coverage docs (pointed to from the PyPI page).  Nowhere can I verify, starting from PyPI, that `github.com/netdbat` belongs to the author of Py-Coverage.  Moreover, the checksum reported for the `4.3.4` tarball from GitHub is different from that reported on the PyPI page for the same tarball.  **Conclusion**: For all these reasons, please revert `py-coverage` back to using PyPI.  Something like this:

```
    homepage = "http://nedbatchelder.com/code/coverage/"
 
    version('4.3.4', '89759813309185efcf4af8b9f7762630',
        url='https://pypi.python.org/packages/6e/33/01cb50da2d0582c077299651038371dba988248058e03c7a7c4be0c84c40/coverage-4.3.4.tar.gz')
     version('4.0a6', '1bb4058062646148965bef0796b61efc',
        url='https://pypi.python.org/packages/source/c/coverage/coverage-4.0a6.tar.gz')
```
Someone should also verify that the URL for `4.0a6` is still valid.  If not ,feel free to just drop the version rather than discovering the new URL at PyPI.

citibeth commented on var/spack/repos/builtin/packages/py-flask/package.py:
```diff
@@ -0,0 +1,41 @@
+    """A microframework based on Werkzeug, Jinja2 and good intentions"""
+
+    homepage = "http://github.com/pallets/flask"
+    url      = "https://github.com/pallets/flask/archive/0.12.tar.gz"
```
Please remove this `url`, it is not needed.

citibeth commented on var/spack/repos/builtin/packages/py-flask/package.py:
```diff
@@ -0,0 +1,41 @@
+    depends_on('py-werkzeug@0.7:',      type=('build', 'run'))
+    depends_on('py-jinja2@2.4:',        type=('build', 'run'))
+    depends_on('py-itsdangerous@0.21:', type=('build', 'run'))
+    depends_on('py-click@2.0:',         type=('build', 'run'))
```
Should these version restrictions all have `when=@0.11:` on them?

healther replied:
I'm not sure how to handle these: I know that 0.11 depends on them, but at least some are older -> should I do the minimal stuff, i.e. assume that the previous versions didn't depend on them (which will fail in some cases) or go the maximal way and introduce dependencies which aren't necessary. 
I personally would prefer unnecessary dependencies but working packages over non-working packages.

adamjstewart replied:
@citibeth Why would these have version restrictions? @healther I think this is fine as is. There is only one version in the package, and so we only claim to support one version. If someone needs to add an older version for whatever reason, it is on them to confirm whether or not the dependencies are required. I see no reason to limit when the dependencies are added.

citibeth commented on var/spack/repos/builtin/packages/py-httpbin/package.py:
```diff
@@ -0,0 +1,42 @@
+    depends_on('py-flask@0.10.1:',      type=('build', 'run'))
+    depends_on('py-itsdangerous@0.24:', type=('build', 'run'))
+    depends_on('py-markupsafe@0.23:',   type=('build', 'run'))
+    depends_on('py-six@1.6.1:',         type=('build', 'run'))
```
Should these version restrictions all have `when=@0.11:` on them?

healther replied:
Same story as above: It depends: Do you prefer no unnecessary dependencies or non-working packages?

citibeth commented on var/spack/repos/builtin/packages/py-pbr/package.py:
```diff
@@ -34,3 +34,4 @@ class PyPbr(PythonPackage):
     version('1.8.1', 'c8f9285e1a4ca6f9654c529b158baa3a')
 
     depends_on('py-setuptools', type='build')
+    depends_on('py-enum34', type='build')
```
It looks like `py-pbr` does not work on all versions of Python; or at least, `py-enum34` is only required for certain older Python versions.  Please indicate --- not just for this package but for all --- what versions of Python it works with.  If it is Python2-only, add `depends_on('python@2:2.8')`.  Otherwise, add a comment instead saying `# This works with Python2 and Python3`.

healther replied:
At least for python 2.7.13 and 3.3.6 ```enum``` is necessary. For 3.4.3 ```enum``` is no longer necessary. I don't know how to put this into the ```depends_on```-syntax, can you help me out?

adamjstewart replied:
That would look like:
```python 
depends_on('enum34', when='^python@:3.4.2')
```
However, I would highly recommend against using that. Concretization is currently very broken, and this exact syntax has caused me woes in the past. See `py-flake8` and dependencies where I had to comment this out to get it to work.

healther replied:
I added a comment, but that is not particularly helpful -> Please advise on how this should be handled (where can I put a guarding ```if``` clause to protect against that failing silently?)

citibeth commented on var/spack/repos/builtin/packages/py-pysocks/package.py:
```diff
@@ -0,0 +1,37 @@
+    """A Python SOCKS client module."""
+
+    homepage = "https://github.com/Anorov/PySocks"
+    url      = "https://github.com/Anorov/PySocks/archive/1.6.6.tar.gz"
```
Remove the URL, it is uselss here.

healther replied:
It offers not checksummed versions of PySocks -> not necessarily that useful, but I'd prefer to have it.

adamjstewart replied:
Can you convert to `pypi.io` then? Otherwise my url parsing algorithm will think it was broken.

citibeth commented on var/spack/repos/builtin/packages/py-itsdangerous/package.py:
```diff
@@ -0,0 +1,37 @@
+    """Various helpers to pass trusted data to untrusted environments."""
+
+    homepage = "http://github.com/mitsuhiko/itsdangerous"
+    url = "https://github.com/pallets/itsdangerous/archive/0.24.tar.gz"
```
Remove the URL, it is useless here.

healther replied:
See above

citibeth commented on var/spack/repos/builtin/packages/py-httpbin/package.py:
```diff
@@ -0,0 +1,42 @@
+    """HTTP Request and Response Service"""
+
+    homepage = "https://github.com/Runscope/httpbin"
+    url      = "https://github.com/Runscope/httpbin/archive/v0.5.0.tar.gz"
```
Remove the URL line.

citibeth commented on var/spack/repos/builtin/packages/py-enum34/package.py:
```diff
@@ -29,7 +29,6 @@ class PyEnum34(PythonPackage):
     """Python 3.4 Enum backported to 3.3, 3.2, 3.1, 2.7, 2.6, 2.5, and 2.4."""
```
This package needs the line:
```
depends_on('python@2:3.3')
```
@adamjstewart did I get the syntax right for that?

healther replied:
I'm not sure if that's what we want, because ```py-enum34``` should not be installed at all for ```python@3.4:```, it's more a "only install for" instead of a "if installed install also", or am I missing something?

citibeth replied:
You need two things: (1) `py-enum34` does not work with `python@3.4:`, so that should be stated in `py-enump34`.  (2) In packages that depend on `py-enump34`, you should make the `depends_on()` be conditional on the Python version.

citibeth commented on var/spack/repos/builtin/packages/py-click/package.py:
```diff
@@ -0,0 +1,37 @@
+    """A simple wrapper around optparse for powerful command line utilities."""
+
+    homepage = "http://github.com/mitsuhiko/click"
+    url = "https://github.com/pallets/click/archive/6.7.tar.gz"
```
Remove this URL

citibeth commented on var/spack/repos/builtin/packages/git-review/package.py:
```diff
@@ -0,0 +1,50 @@
+    """git-review is a tool that helps submitting git branches to gerrit"""
+
+    homepage = "http://docs.openstack.org/infra/git-review"
+    url = "https://github.com/openstack-infra/git-review/archive/1.25.0.tar.gz"
```
Remove this URL

citibeth commented on var/spack/repos/builtin/packages/git-review/package.py:
```diff
@@ -0,0 +1,50 @@
+
+    version('1.25.0', '0a061d0e23ee9b93c6212a3fe68fb7ab',
+        url='https://pypi.python.org/packages/92/38/b97fee6752540a92f44e405a51c53d3c36b5d57e139683862597b09a5c6c/git-review-1.25.0.tar.gz')
+    version('1.24',   '145116fe58a3487c3ad1bf55538fd741')
```
Add PyPI URLs for old versions.

citibeth commented on var/spack/repos/builtin/packages/py-pytest-cov/package.py:
```diff
@@ -0,0 +1,41 @@
+    version('2.3.1', '8e7475454313a035d08f387ee6d725cb',
+        url='https://pypi.python.org/packages/4c/2f/de9e8f226d4fb86f330ee2fcc709204fbd3aeab6ce17756b3cf5ea9aa4d7/pytest-cov-2.3.1.tar.gz')
+
+    extends('python', ignore=r'bin/*')
```
Can you please add a comment explaining this `ignore=` clause?  It's not commonly used in Spack, and deserves a bit of explanation.

citibeth commented on var/spack/repos/builtin/packages/py-pytest-cov/package.py:
```diff
@@ -0,0 +1,41 @@
+    """Pytest plugin for measuring coverage."""
+
+    homepage = "https://github.com/pytest-dev/pytest-cov"
+    url      = "https://github.com/pytest-dev/pytest-cov/archive/v2.4.0.tar.gz"
```
Remove the URL

citibeth commented on var/spack/repos/builtin/packages/py-pbr/package.py:
```diff
@@ -34,3 +34,4 @@ class PyPbr(PythonPackage):
     version('1.8.1', 'c8f9285e1a4ca6f9654c529b158baa3a')
```
Convert to PyPI URL

citibeth commented on var/spack/repos/builtin/packages/py-pytest-httpbin/package.py:
```diff
@@ -0,0 +1,44 @@
+    version('0.2.3', 'b8ebb8e2fbac1a445fb5d044f7fec556',
+        url='https://pypi.python.org/packages/12/12/5430600cb9417080b561237761ff2dffde520b664cc352433d2e57051222/pytest-httpbin-0.2.3.tar.gz')
+
+    extends('python', ignore=r'bin/flask')
```
Can you please add a comment explaining this `ignore=` clause?  It's not commonly used in Spack, and deserves a bit of explanation.

citibeth commented on var/spack/repos/builtin/packages/py-pytest-httpbin/package.py:
```diff
@@ -0,0 +1,44 @@
+    """Easily test your HTTP library against a local copy of httpbin"""
+
+    homepage = "https://github.com/kevin1024/pytest-httpbin"
+    url      = "https://github.com/kevin1024/pytest-httpbin/archive/v0.2.0.tar.gz"
```
Remove URL

citibeth commented on var/spack/repos/builtin/packages/py-pytest-mock/package.py:
```diff
@@ -0,0 +1,41 @@
+    version('1.2', 'a7fa820f7bc71698660945836ff93c73',
+        url="https://pypi.python.org/packages/30/11/a5a8009eff04bc15c37e2f8e33d8ed99adf828ec8f551fb31d99f6c73b5b/pytest-mock-1.2.zip")
+
+    extends('python', ignore=r'bin/*')
```
Can you please add a comment explaining this `ignore=` clause?  It's not commonly used in Spack, and deserves a bit of explanation.

citibeth commented on var/spack/repos/builtin/packages/py-pytest-mock/package.py:
```diff
@@ -0,0 +1,41 @@
+    """Thin-wrapper around the mock package for easier use with py.test"""
+
+    homepage = "https://github.com/pytest-dev/pytest-mock"
+    url      = "https://github.com/pytest-dev/pytest-mock/archive/v1.5.0.tar.gz"
```
Remove URL

citibeth commented on var/spack/repos/builtin/packages/py-requests/package.py:
```diff
@@ -0,0 +1,61 @@
+    """Python HTTP for Humans."""
+
+    homepage = "http://python-requests.org"
+    url = "https://github.com/kennethreitz/requests/archive/v2.13.0.tar.gz"
```
Remove this URL

citibeth commented on var/spack/repos/builtin/packages/py-requests/package.py:
```diff
@@ -0,0 +1,61 @@
+    homepage = "http://python-requests.org"
+    url = "https://github.com/kennethreitz/requests/archive/v2.13.0.tar.gz"
+
+    version('2.13.0', '94ad79c2e57917aca999308b1fb4cbb4')
```
Look here for PyPI URLs: https://pypi.python.org/pypi/requests/2.13.0#downloads

citibeth commented on var/spack/repos/builtin/packages/py-werkzeug/package.py:
```diff
@@ -0,0 +1,37 @@
+    """The Swiss Army knife of Python web development"""
+
+    homepage = "http://werkzeug.pocoo.org"
+    url      = "https://github.com/pallets/werkzeug/archive/0.11.15.tar.gz"
```
Remove URL

citibeth commented on var/spack/repos/builtin/packages/git-review/package.py:
```diff
@@ -0,0 +1,50 @@
+class GitReview(PythonPackage):
+    """git-review is a tool that helps submitting git branches to gerrit"""
+
+    homepage = "http://docs.openstack.org/infra/git-review"
```
Please rename this package to `py-git-review/package.yaml`.

citibeth commented:
For finding old versions off of PyPI, please use the script below.  For example, `python ver.py git-release` yields the following:
```
$ python ./ver.py git-review
    version('1.25.0', '0a061d0e23ee9b93c6212a3fe68fb7ab',
        url='https://pypi.python.org/packages/92/38/b97fee6752540a92f44e405a51c53d3c36b5d57e139683862597b09a5c6c/git-review-1.25.0.tar.gz'
    version('1.24', '537c78591e74853c204c5b3d88c0c4fd',
        url='https://pypi.python.org/packages/40/c9/c007572552b78cfe8836cd8a836d252d6b2c0e8dc53b3689b4e243d0a780/git-review-1.24.tar.gz'
    version('1.23', '1b3414179d27b2e93615befe118809cb',
        url='https://pypi.python.org/packages/b7/05/9adc7ac55540f926fe8b80ff1d2e26c0bd1bdd57ca13f258fe1bbe5aa035/git-review-1.23.tar.gz'
...
```
Here is the source:
```
from __future__ import print_function
import xmlrpclib
import os
import sys

# https://wiki.python.org/moin/PyPIXmlRpc

ext_sortorder = {'.tar.gz':0, '.zip':1}

depends_on_template = "    depends_on('{package}', when='@{version}')"

version_template = \
"""    version('{version}', '{md5_digest}',
        url='{url}')"""


package_template = """
from spack import *
import platform
class {package}(PyPIPackage):
    \"\"\"{description}\"\"\"
    homepage = "{home_page}"
{versions}
{depends_ons}
    extends('python')
"""




def get_source_url(name, version):
    """Returns the URL needed to download a package from PyPI.
    name:
        Name of the package in PyPI
    version: (str)
        Version of the pacakge to download
    extension:
        Filename extension desired
    Raises KeyError if the desired PyPI URL cannot be located.
    See: https://wiki.python.org/moin/PyPIXmlRpc
    """

    client = xmlrpclib.ServerProxy('https://pypi.python.org/pypi')
    releases = client.release_urls(name, version)

    good_url_recs = []
    for release in releases:
        if release['packagetype'] != 'sdist':
            continue
        _,ext = os.path.splitext(release['filename'])
        good_url_recs.append((ext_sortorder.get(ext,len(ext_sortorder)), release))

    good_url_recs.sort()
    if len(good_url_recs) > 0:
        return good_url_recs[0][1]
    else:
        return None

def get_versions(pypi, package):
    # Get all version info for this package
    versions = []
    for ver in pypi.package_releases(package, True):
        data = pypi.release_data(package, ver)

        # Not all package versions have a valid URL that is useful to us.
        url = get_source_url(package,ver)
        if url is not None:
            data.update(url)
        else:
            data['url'] = data['download_url']
            data['md5_digest'] = ''    # TODO: Download and figure this out
        yield data


def write_package_list(pypi):
    with open('pypi-packages.txt', 'w') as fout:
        for package in pypi.list_packages():
            fout.write(package + '\n')


def convert_package(pypi, package):
    versions = get_versions(pypi, package)
    version_lines = []
    depends_on_lines = []
    for ver in versions:
        version_lines.append(version_template.format(
            version=ver['version'],
            md5_digest=ver['md5_digest'],
            url=ver['url']))
        for requires in ver.get('requires_dist', []):
            depends_on_lines.append(depends_on_template.format(
                package=requires,
                version=ver['version']))

    ver0 = versions[0]

    spack_package = package_template.format(
        package = ver0['name'],
        description = '\n    '.join(ver0['description'].splitlines()),
        home_page = ver0['home_page'],
        versions = '\n'.join(version_lines),
        depends_ons = '\n'.join(depends_on_lines)
    )

    print(spack_package)


def dump_package(pypi, package):
    versions = get_versions(pypi, package)
    print(package)
    for ver in versions:
        print('  {0}'.format(ver['version']))
        for k,v in ver.items():
            print('    {0}: {1}'.format(k,v))

def main():
    pkg = sys.argv[1]
    pypi = xmlrpclib.ServerProxy('https://pypi.python.org/pypi')


    for ver in get_versions(pypi, pkg):
        keys = {'version' : ver['version']}
        keys['url'] = '<FIXME>'
        keys['md5'] = '<FIXME>'
        try:
            src_url = get_source_url(pkg, ver['version'])
            keys['url'] = src_url['url']
            keys['md5'] = src_url['md5_digest']
        except:
            pass

        print("    version('{version}', '{md5}',\n" \
              "        url='{url}'".format(**keys))

#    write_package_list(pypi)
#    walk_pypi(pypi)
#    convert_package(pypi, sys.argv[1])
#    dump_package(pypi, 'basemap')

main()
```

adamjstewart commented:
> For finding old versions off of PyPI, please use the script below. For example, `python ver.py git-release` yields the following:
> 
> ```
> $ python ./ver.py git-review
>     version('1.25.0', '0a061d0e23ee9b93c6212a3fe68fb7ab',
>         url='https://pypi.python.org/packages/92/38/b97fee6752540a92f44e405a51c53d3c36b5d57e139683862597b09a5c6c/git-review-1.25.0.tar.gz'
>     version('1.24', '537c78591e74853c204c5b3d88c0c4fd',
>         url='https://pypi.python.org/packages/40/c9/c007572552b78cfe8836cd8a836d252d6b2c0e8dc53b3689b4e243d0a780/git-review-1.24.tar.gz'
>     version('1.23', '1b3414179d27b2e93615befe118809cb',
>         url='https://pypi.python.org/packages/b7/05/9adc7ac55540f926fe8b80ff1d2e26c0bd1bdd57ca13f258fe1bbe5aa035/git-review-1.23.tar.gz'
> ```

Alternatively, use the `pypi.io` URL:
```python
url = "https://pypi.io/packages/source/g/git-review/git-review-1.25.0.tar.gz"

version('1.25.0', '0a061d0e23ee9b93c6212a3fe68fb7ab')
version('1.24',   '537c78591e74853c204c5b3d88c0c4fd')
version('1.23',   '1b3414179d27b2e93615befe118809cb')
```

citibeth commented:
I see.  `pypi.io` looks like it's bringing sanity back to PyPI URLs.  Are
all PyPI packages up on it, or is it hit-and-miss?

On Tue, Feb 14, 2017 at 10:40 AM, Adam J. Stewart <notifications@github.com>
wrote:

> For finding old versions off of PyPI, please use the script below. For
> example, python ver.py git-release yields the following:
>
> $ python ./ver.py git-review
>     version('1.25.0', '0a061d0e23ee9b93c6212a3fe68fb7ab',
>         url='https://pypi.python.org/packages/92/38/b97fee6752540a92f44e405a51c53d3c36b5d57e139683862597b09a5c6c/git-review-1.25.0.tar.gz'
>     version('1.24', '537c78591e74853c204c5b3d88c0c4fd',
>         url='https://pypi.python.org/packages/40/c9/c007572552b78cfe8836cd8a836d252d6b2c0e8dc53b3689b4e243d0a780/git-review-1.24.tar.gz'
>     version('1.23', '1b3414179d27b2e93615befe118809cb',
>         url='https://pypi.python.org/packages/b7/05/9adc7ac55540f926fe8b80ff1d2e26c0bd1bdd57ca13f258fe1bbe5aa035/git-review-1.23.tar.gz'
>
> Alternatively, use the pypi.io URL:
>
> url = "https://pypi.io/packages/source/g/git-review/git-review-1.25.0.tar.gz"
>
> version('1.25.0', '0a061d0e23ee9b93c6212a3fe68fb7ab')
> version('1.24',   '537c78591e74853c204c5b3d88c0c4fd')
> version('1.23',   '1b3414179d27b2e93615befe118809cb')
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3088#issuecomment-279742759>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd7MsVMWlhdMjqVs35AmcTu8d2jmYks5rccr4gaJpZM4L75Wy>
> .
>

adamjstewart commented:
@citibeth All that I've come across. We've had a few discussions about this in the past.

healther commented:
Okay so in general ```spack``` prefers the URLs ```pip``` would look at for python packages right? Then I'll move the main URLs to `pypi.io`, but I'd actually also give the `github.com/.../archive/...` URLs, they automatically add additional versions (even though they are not checksummed).

@citibeth If you agree to that as a general policy I'd do that later, if you disagree, please explain, what the policy is/should be and I'll adapt to that. Again I don't know the implications and as such don't have a strong opinion, but would prefer of only doing the work once

citibeth commented:
@adamjstewart what is your final thought on this?
On Feb 16, 2017 7:17 AM, "healther" <notifications@github.com> wrote:

> Okay so in general spack prefers the URLs pip would look at for python
> packages right? Then I'll move the main URLs to pypi.io, but I'd actually
> also give the github.com/.../archive/... URLs, they automatically add
> additional versions (even though they are not checksummed).
>
> @citibeth <https://github.com/citibeth> If you agree to that as a general
> policy I'd do that later, if you disagree, please explain, what the policy
> is/should be and I'll adapt to that. Again I don't know the implications
> and as such don't have a strong opinion, but would prefer of only doing the
> work once
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3088#issuecomment-280315661>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd4i5b5D51-1eM4WJ9f1V7nAFWqtJks5rdD5MgaJpZM4L75Wy>
> .
>

adamjstewart commented:
I'd prefer only using `pypi.io`. GitHub may have additional versions, but there is probably a reason that the developer didn't put those versions on PyPI. They may have a bug, or be development releases for all we know.

healther commented:
That should take care of most of the url issues. I tested most urls if I copy/paste them into a browser. I did not check the checksums against the spack implemented ones.

I also noted that for the packages where I tested it
```spack versions <name>```
never returned any unchecked checksums, this makes me a bit skeptical. Can you please do a sanity check?

adamjstewart commented on var/spack/repos/builtin/packages/py-enum34/package.py:
```diff
@@ -29,15 +29,13 @@ class PyEnum34(PythonPackage):
     """Python 3.4 Enum backported to 3.3, 3.2, 3.1, 2.7, 2.6, 2.5, and 2.4."""
 
     homepage = "https://pypi.python.org/pypi/enum34"
-    url      = "https://pypi.python.org/packages/source/e/enum34/enum34-1.1.6.tar.gz"
```
Can you convert this package to use `pypi.io`? I confirmed that it works. Also remove the URL from the version directive below.

healther replied:
Done.
But I'm a bit confused atm, what is the desired default for a new package: Having a general ```url=...``` or one for each version?

adamjstewart replied:
The desired default is that every package has a single `url` declared at the package level. Obviously, this isn't possible for all packages, as some versions have a single weird URL that cannot be extrapolated. For Python packages, we should be able to get everything from `pypi.io`, which makes life a lot easier. Otherwise, every time you add a new version, you have to add a `url` for that version, and it gets more complicated than necessary.

adamjstewart commented on var/spack/repos/builtin/packages/py-enum34/package.py:
```diff
@@ -29,15 +29,13 @@ class PyEnum34(PythonPackage):
             url="https://pypi.python.org/packages/bf/3e/31d502c25302814a7c2f1d3959d2a3b3f78e509002ba91aea64993936876/enum34-1.1.6.tar.gz")
 
-    depends_on('python@2.4:2.8,3.3:')
+    depends_on('python@2.4:3.3.99')
```
`2.4:3.3` is fine here. But yes, you are right, it should work for Python 3.1 and 3.2. I was confused when I added this line because the `setup.py` contains:
```python
        classifiers=[                                                           
            'Development Status :: 5 - Production/Stable',                      
            'Intended Audience :: Developers',                                  
            'License :: OSI Approved :: BSD License',                           
            'Programming Language :: Python',                                   
            'Topic :: Software Development',                                    
            'Programming Language :: Python :: 2.4',                            
            'Programming Language :: Python :: 2.5',                            
            'Programming Language :: Python :: 2.6',                            
            'Programming Language :: Python :: 2.7',                            
            'Programming Language :: Python :: 3.3',                            
            'Programming Language :: Python :: 3.4',                            
            'Programming Language :: Python :: 3.5',                            
            ],                                     
```
At this point I'm not sure what this means anymore. I thought it meant that Python 3.0 - 3.2 were not supported, but I guess not.

healther replied:
I don't explicitly know. I confirmed once that it builds with 3.2.6, but nothing else. According to its description it should already be included in the standard installation of 3.4+, nevertheless it builds there...

However I don't have access to this system right now and only my home system (bash on windows) and here python 3.2. doesn't build at all... I hope that our ```afs``` is fixed tomorrow, then I can confirm that at work.

A more general question: Does ```:3.3``` include ```3.3.1``` or does it not?

adamjstewart replied:
We may want to remove the version constraints altogether then. Just because it isn't needed in newer versions of Python doesn't mean that someone might not want to test a pre-release newer than what comes with python.

Yes, `:3.3` should include `3.3.1`. I'm away from my computer right now, but I'll verify this later. If not it's a bug :smiley:

adamjstewart replied:
> Does `:3.3` include `3.3.1` or does it not?

Did some quick tests to confirm this. Version ranges are _inclusive_:
```
$ spack python
Spack version 0.10.0
Python 2.7.12, Darwin x86_64
>>> from spack.version import *
>>> Version('2.3') in VersionRange('2.4', '3.3')
False
>>> Version('2.4') in VersionRange('2.4', '3.3')
True
>>> Version('3.3') in VersionRange('2.4', '3.3')
True
>>> Version('3.4') in VersionRange('2.4', '3.3')
False
```
and _include minor versions_ as well:
```
>>> Version('2.3.9') in VersionRange('2.4', '3.3')
False
>>> Version('2.4.1') in VersionRange('2.4', '3.3')
True
>>> Version('3.3.1') in VersionRange('2.4', '3.3')
True
>>> Version('3.4.1') in VersionRange('2.4', '3.3')
False
```

adamjstewart commented on var/spack/repos/builtin/packages/py-git-review/package.py:
```diff
@@ -0,0 +1,49 @@
+    depends_on('tk',               type=('run'))
+
+    def setup_environment(self, spack_env, run_env):
+        run_env.set('PBR_VERSION', self.spec.version)
```
This sets `PBR_VERSION` to the version of `py-git-review`. Just to make sure, is this what you intended?

adamjstewart commented on var/spack/repos/builtin/packages/py-itsdangerous/package.py:
```diff
@@ -0,0 +1,36 @@
+    """Various helpers to pass trusted data to untrusted environments."""
+
+    homepage = "http://github.com/mitsuhiko/itsdangerous"
+    url = "https://pypi.io/packages/source/i/itsdangerous/itsdangerous-1.5.7.tar.gz"
```
Shouldn't the version in the URL be `0.24`?

healther replied:
spack is smart enough, but I can change it, if you like
```
spack info py-itsdangerous
PythonPackage:    py-itsdangerous
Homepage:         http://github.com/mitsuhiko/itsdangerous

Safe versions:
    0.24    https://pypi.io/packages/source/i/itsdangerous/itsdangerous-0.24.tar.gz

Variants:
    None

Installation Phases:
    build    install

Build Dependencies:
    py-setuptools  python

Link Dependencies:
    python

Run Dependencies:
    None

Virtual Packages:
    None

Description:
    Various helpers to pass trusted data to untrusted environments.
```

adamjstewart replied:
Please do. I'm working on our automatic version detection right now and it relies on the detected version to be present in the package in order to verify that it was parsed correctly.

adamjstewart commented on var/spack/repos/builtin/packages/py-pysocks/package.py:
```diff
@@ -0,0 +1,37 @@
+    """A Python SOCKS client module."""
+
+    homepage = "https://github.com/Anorov/PySocks"
+    url      = "https://pypi.io/packages/source/s/socks/socks-1.5.7.tar.gz"
```
This URL doesn't work. The correct URL is:
https://pypi.io/packages/source/P/PySocks/PySocks-1.6.6.tar.gz

adamjstewart replied:
Ping @healther. The correct URL should be https://pypi.io/packages/source/P/PySocks/PySocks-1.6.6.tar.gz

adamjstewart commented on var/spack/repos/builtin/packages/py-pytest-cov/package.py:
```diff
@@ -0,0 +1,40 @@
+    homepage = "https://github.com/pytest-dev/pytest-cov"
+
+    version('2.3.1', '8e7475454313a035d08f387ee6d725cb',
+            url='https://pypi.python.org/packages/4c/2f/de9e8f226d4fb86f330ee2fcc709204fbd3aeab6ce17756b3cf5ea9aa4d7/pytest-cov-2.3.1.tar.gz')
```
Can you convert this to `pypi.io`? The URL is:
https://pypi.io/packages/source/p/pytest-cov/pytest-cov-2.3.1.tar.gz

adamjstewart commented on var/spack/repos/builtin/packages/py-pytest-httpbin/package.py:
```diff
@@ -0,0 +1,43 @@
+    homepage = "https://github.com/kevin1024/pytest-httpbin"
+
+    version('0.2.3', 'b8ebb8e2fbac1a445fb5d044f7fec556',
+            url='https://pypi.python.org/packages/12/12/5430600cb9417080b561237761ff2dffde520b664cc352433d2e57051222/pytest-httpbin-0.2.3.tar.gz')
```
`pypi.io`:
https://pypi.io/packages/source/p/pytest-httpbin/pytest-httpbin-0.2.3.tar.gz

adamjstewart commented on var/spack/repos/builtin/packages/py-pytest-httpbin/package.py:
```diff
@@ -0,0 +1,43 @@
+    extends('python', ignore=r'bin/flask')
+
+    depends_on('py-setuptools', type='build')
+    depends_on('py-pytest',     type=('build', 'run'))
```
Does this actually depend on `py-pytest`? I don't see anything in the `setup.py` that suggests that.

adamjstewart commented on var/spack/repos/builtin/packages/py-pytest-mock/package.py:
```diff
@@ -0,0 +1,40 @@
+    homepage = "https://github.com/pytest-dev/pytest-mock"
+
+    version('1.2', 'a7fa820f7bc71698660945836ff93c73',
+            url="https://pypi.python.org/packages/30/11/a5a8009eff04bc15c37e2f8e33d8ed99adf828ec8f551fb31d99f6c73b5b/pytest-mock-1.2.zip")
```
`pypi.io`:
https://pypi.io/packages/source/p/pytest-mock/pytest-mock-1.2.zip

adamjstewart commented on var/spack/repos/builtin/packages/py-requests/package.py:
```diff
@@ -0,0 +1,61 @@
+    """Python HTTP for Humans."""
+
+    homepage = "http://python-requests.org"
+    url = "https://pypi.io/packages/source/r/requests/requests-1.25.0.tar.gz"
```
Can you use `2.13.0` in this URL? `1.25.0` doesn't exist.

adamjstewart commented on var/spack/repos/builtin/packages/py-requests/package.py:
```diff
@@ -0,0 +1,61 @@
+    depends_on('py-sphinx@1.3.5:',           type=('build', 'run'))
+    depends_on('py-sphinx-rtd-theme@0.1.9:', type=('build', 'run'))
+    depends_on('py-werkzeug@0.11.4:',        type=('build', 'run'))
+    depends_on('py-wheel@0.29.0:',           type=('build', 'run'))
```
I'm not seeing any dependencies in the `setup.py`, but I'm sure there are some. Where did you get this list from?

I'm seeing the following test requirements:
```python
test_requirements = ['pytest>=2.8.0', 'pytest-httpbin==0.0.7', 'pytest-cov', 'pytest-mock']
```
These shouldn't be added as dependencies until #1279 and #2869 are complete.

healther replied:
I have no idea... I confirmed that it builds with only ```py-setuptools``` but I'll ask around.

adamjstewart replied:
When you build with only setuptools, check to see if pip automatically installs any dependencies. Also check the site-packages directory and see if any other things got installed. Those would be required dependencies.

adamjstewart commented on var/spack/repos/builtin/packages/py-werkzeug/package.py:
```diff
@@ -0,0 +1,36 @@
+    """The Swiss Army knife of Python web development"""
+
+    homepage = "http://werkzeug.pocoo.org"
+    url      = "https://pypi.io/packages/source/w/werkzeug/werkzeug-0.11.11.tar.gz"
```
This URL should be:
https://pypi.io/packages/source/W/Werkzeug/Werkzeug-0.11.11.tar.gz

Btw, 0.11.15 is out now if you are interested. You don't have to add it though.

healther replied:
why should it be a capital W? is that important? Or am I missing something?

adamjstewart replied:
I tried fetching with wget with the lowercase W and it didn't work. Capital W did.

adamjstewart commented on var/spack/repos/builtin/packages/py-wheel/package.py:
```diff
@@ -29,8 +29,10 @@ class PyWheel(PythonPackage):
+    version('0.30.0', 'aa1145d48e4da2f9415eac2d51468a83',
+        url='https://pypi.python.org/packages/a7/37/947b4329c4a3c72093b6c8e9b4be8c7f10c32dbb78848d3a234ce01c059d/wheel-0.30.0a0.tar.gz')
+    version('0.26.0', '4cfc6e7e3dc7377d0164914623922a10',
+        url='https://pypi.python.org/packages/source/w/wheel/wheel-0.26.0.tar.gz')
```
Can you use `pypi.io` for this package?:
https://pypi.io/packages/source/w/wheel/wheel-0.30.0a0.tar.gz

Also, the version should be `0.30.0a0`, not `0.30.0`.

citibeth commented:
Is this unit tested?  I think it is...
On Feb 20, 2017 6:26 PM, "Adam J. Stewart" <notifications@github.com> wrote:

> *@adamjstewart* commented on this pull request.
> ------------------------------
>
> In var/spack/repos/builtin/packages/py-enum34/package.py:
>
> >
>      version('1.1.6', '5f13a0841a61f7fc295c514490d120d0',
>              url="https://pypi.python.org/packages/bf/3e/31d502c25302814a7c2f1d3959d2a3b3f78e509002ba91aea64993936876/enum34-1.1.6.tar.gz")
>
> -    depends_on('python@2.4:2.8,3.3:')
> +    depends_on('python@2.4:3.3.99')
>
> Does :3.3 include 3.3.1 or does it not?
>
> Did some quick tests to confirm this. Version ranges are *inclusive*:
>
> $ spack python
> Spack version 0.10.0
> Python 2.7.12, Darwin x86_64
> >>> from spack.version import *
> >>> Version('2.3') in VersionRange('2.4', '3.3')
> False
> >>> Version('2.4') in VersionRange('2.4', '3.3')
> True
> >>> Version('3.3') in VersionRange('2.4', '3.3')
> True
> >>> Version('3.4') in VersionRange('2.4', '3.3')
> False
>
> and *include minor versions* as well:
>
> >>> Version('2.3.9') in VersionRange('2.4', '3.3')
> False
> >>> Version('2.4.1') in VersionRange('2.4', '3.3')
> True
> >>> Version('3.3.1') in VersionRange('2.4', '3.3')
> True
> >>> Version('3.4.1') in VersionRange('2.4', '3.3')
> False
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3088#discussion_r102107188>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cdxJUXiJ1PyqeArS0vMdyh1tLL05Sks5reiEdgaJpZM4L75Wy>
> .
>

adamjstewart commented:
@citibeth What do you mean by unit tested? We don't build any packages during unit tests, if that's what you mean.

citibeth commented:
I mean the veosion matching stuff you just tested manually.
On Feb 20, 2017 6:51 PM, "Adam J. Stewart" <notifications@github.com> wrote:

> @citibeth <https://github.com/citibeth> What do you mean by unit tested?
> We don't build any packages during unit tests, if that's what you mean.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/LLNL/spack/pull/3088#issuecomment-281208118>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AB1cd-CZC_lCgT76ewirtYheRmNdAn6Eks5reicXgaJpZM4L75Wy>
> .
>

adamjstewart commented:
Good question! It is actually:
https://github.com/LLNL/spack/blob/develop/lib/spack/spack/test/versions.py#L201

Our unit tests look pretty good, although there are some oddities I've discovered that we don't seem to be testing:
```
$ spack python
Spack version 0.10.0
Python 2.7.12, Darwin x86_64
>>> from spack.version import *
>>> Version('1.2.3') == Version('1_2_3') == Version('1-2-3')
True
```

healther commented:
Not tested in the sense that I couldn't try to build it yet (our system has some filesystem problems atm...) So if one of you can do a ```spack install py-git-review``` and it works w/o problems: LGTM. Otherwise: I can't test it myself atm. Will do when the system is back up

tgamblin commented:
@healther: I think @adamjstewart had a couple more questions on the package; if those are addressed can you say so (not sure if your last comment is saying so).

@adamjstewart @citibeth: does this look ok to you now?

citibeth commented on var/spack/repos/builtin/packages/py-pbr/package.py:
```diff
@@ -29,8 +29,12 @@ class PyPbr(PythonPackage):
     version('1.8.1', 'c8f9285e1a4ca6f9654c529b158baa3a')
 
     depends_on('py-setuptools', type='build')
+    # Only needed for py<3.4, however when='^python@:3.4.2' syntax is broken
```
@tgamblin Can you look at this?  How do we depend on `py-enum34` for only certain versions of Python?

adamjstewart replied:
We can't until the concretizer is fixed. I've ran into this problem before myself. In the meantime, it doesn't hurt to build it for a version of Python that already has it.

healther commented:
Sorry for the delay... This week is absolutely crazy...
@adamjstewart I fixed the ```py-pysocks``` URL and build it (that's what took me so long to get to). It builds without errors, but I haven't tested whether there are actually unneeded dependencies.

adamjstewart commented:
@healther 
```
$ wget https://pypi.io/packages/source/s/socks/socks-1.6.6.tar.gz
--2017-02-23 14:42:17--  https://pypi.io/packages/source/s/socks/socks-1.6.6.tar.gz
Resolving pypi.io... 151.101.128.223, 151.101.192.223, 151.101.0.223, ...
Connecting to pypi.io|151.101.128.223|:443... connected.
HTTP request sent, awaiting response... 301 Redirect to Primary Domain
Location: https://pypi.org/packages/source/s/socks/socks-1.6.6.tar.gz [following]
--2017-02-23 14:42:17--  https://pypi.org/packages/source/s/socks/socks-1.6.6.tar.gz
Resolving pypi.org... 151.101.64.223, 151.101.192.223, 151.101.128.223, ...
Connecting to pypi.org|151.101.64.223|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://files.pythonhosted.org/packages/source/s/socks/socks-1.6.6.tar.gz [following]
--2017-02-23 14:42:17--  https://files.pythonhosted.org/packages/source/s/socks/socks-1.6.6.tar.gz
Resolving files.pythonhosted.org... 151.101.57.63, 2a04:4e42:e::319
Connecting to files.pythonhosted.org|151.101.57.63|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2017-02-23 14:42:18 ERROR 404: Not Found.
```
This URL does not exist. The correct URL is:
https://pypi.io/packages/source/P/PySocks/PySocks-1.6.6.tar.gz

`PySocks`, not `socks`.

healther commented:
Okay this time with a working 
```spack install py-pysocks```

I'm not entirely sure why I bundled this package in with the ```py-git-review``` as this doesn't depend on ```py-pysocks``` (hence it didn't break my build this morning)

adamjstewart commented:
@healther Can you resolve the conflicts on this branch?

adamjstewart commented on var/spack/repos/builtin/packages/py-wheel/package.py:
```diff
@@ -29,7 +29,7 @@ class PyWheel(PythonPackage):
 
     homepage = "https://pypi.python.org/pypi/wheel"
-    url      = "https://pypi.python.org/packages/source/w/wheel/wheel-0.26.0.tar.gz"
+    url      = "https://pypi.io/packages/source/w/wheel/wheel-0.30.0a0.tar.gz"
```
Can you make sure the version in the URL is also present in the package?

healther replied:
Done by moving the url to version 0.29.0
